### DISCUSSION TITLE
RAM Loading example code

### DISCUSSION THREAD

**Discussion by Sathishkumar-rk** [mcu-tools/mcuboot#1376] [Q&A]
Dear Team,

I'm new to the MCUBoot project, Is it possible to use the RAM Loading feature in MCUBoot on mimxrt1060-evk? Is there any example code for reference? I was trying to use the MCUBoot with zephyr project.

Regards,
Sathishkumar K

**Comment by d3zd3z** [comment]
I'm not aware of any sample code for it. The RAM loading is fairly new, and at least most of the targets I work with don't have enough RAM for me to really test it.

**Reply by Sathishkumar-rk** [reply]
Can you refer some app note for this feature, it will be helpful to start with app note.

**Reply by Sathishkumar-rk** [reply]
RAM loading option is not showing in Kconfig menu option? Is there any app note or reference note to enable option on KConfig menu

**Reply by Sathishkumar-rk** [reply]
Hi,

Update on the above thread, I went through some branch commits to enable the ram load option and I can able to set the ram load upgrade mode in menu-config using below community thread posts

1. https://github.com/mcu-tools/mcuboot/pull/1336/files/2f9ca5f1b639d5c6d2959eafa4500f1b2a97cb76
2. https://github.com/mcu-tools/mcuboot/issues/1330

Now, the problem is when I'm trying run the application image it throws on error like bus fault error.

error image:
<img width="497" alt="image" src="https://user-images.githubusercontent.com/105291276/179456462-8de1d0ad-d942-4b2a-82e1-f8c6cf9bd2a4.png">

Menu configuration:
<img width="554" alt="image" src="https://user-images.githubusercontent.com/105291276/179456841-54a23e5d-715b-4216-9071-a9a9f6cd884d.png">

Application image preparation:
<img width="859" alt="image" src="https://user-images.githubusercontent.com/105291276/179457055-2e78ba13-24ca-44cd-bd99-d60c26080afa.png">

Hi @danieldegrasse @d3zd3z any thought on above issues?

Thanks,

**Reply by DerekSnell** [reply]
Hi @Sathishkumar-rk ,
I suggest you use a debugger, and step through the MCUboot code to diagnose what is causing this bus fault.  For debugging a Zephyr application on the RT1060, these [Hands-On Labs](https://www.nxp.com/webapp/Download?colCode=LAB-MATERIAL-BUILD-ZEPHYR-APP-MCUS) detail how to use GDB or the MCUXpresso IDE.  Or this [blog ](https://blog.golioth.io/taking-the-next-step-debugging-with-segger-ozone-and-systemview-on-zephyr/) provides detailed steps for using Segger's Ozone debugger.  

Some initial thoughts that come to my mind, your menuconfig settings appear to be copying to SDRAM at 0x8000_2000.  The SEMC needs to be initialized for SDRAM first before accessing the SDRAM.  With the Zephyr applications for RT10xx, this SEMC initialization is typically configured in the DCD, and the ROM bootloader will configure the SDRAM before executing the application.  You should be able to confirm easily if the MCUboot application is able to update the SDRAM.

Also, your bus fault shows an invalid address for the PC causing the fault at 0x0207_F020.  That address does not appear to be valid looking at the RT1060 memory map.  So why is the PC trying to execute from there?  Using a debugger will help you track that down.

Best regards

**Reply by Sathishkumar-rk** [reply]
@DerekSnell  Thanks for update. Yes, you are right problem was when i using sram location before initialization (RT1160 doesn't have the initialization details of sram).

For time being i used DTCM location (0x20004000) instead of sram location (0x80002000), since I'm working sample applications which has low memory consumption. In DTCM location i could able to run the application.

Now, I'm flashed two different version of application. Slot-0 will have lower version and Slot-1 will have the higher version. rsp pointer is getting updated with Slot-1 image but Slot-0 application getting executed, I think need to change something in **do_boot** API in main.c file ? any thought on this?

**Reply by Sathishkumar-rk** [reply]
Hi @DerekSnell @danieldegrasse @d3zd3z 

Please give your thoughts on this issue.

I have attached the main.c file of mcuboot project code.

Is there anything I need to change on do_boot API to execute the latest version of image?

I suspect the new version of image address is updated on **rsp** but while executing Slot-0 image getting started, this make me thing about whether the copied image on RAM is getting executed or is still executing application from Flash itself?

[main.zip](https://github.com/mcu-tools/mcuboot/files/9140605/main.zip)
[config.zip](https://github.com/mcu-tools/mcuboot/files/9140445/config.zip)

Image version of both slot:
<img width="356" alt="Screenshot 2022-07-19 170435" src="https://user-images.githubusercontent.com/105291276/179740785-6e50e3c5-3f8c-4c10-8acd-c8322803f2ad.png">

Ram load information:
<img width="484" alt="Screenshot 2022-07-19 170243" src="https://user-images.githubusercontent.com/105291276/179740530-dfff816b-e447-4599-bd4c-6c7e23e4a35a.png">


Regards,
Sathishkumar K

**Reply by Sathishkumar-rk** [reply]
Hi @DerekSnell @danieldegrasse,

How to initialize the SEMC for accessing SDRAM. I'm using mimxrt1160_evk_cm7 board & zephyr RTOS for my development work.

Your input will be really helpful on developing ram loading upgrade mode.
 
Regards,
Sathishkumar K

**Reply by DerekSnell** [reply]
Linking to your other discussion on this topic https://github.com/zephyrproject-rtos/zephyr/discussions/48242#discussioncomment-3242064
